File: Week3/Code/get_TreeHeight.R
# Takes a file from command line and calculates tree heights.
########################################################################
#                               Functions
########################################################################
# define function TreeHeight.R
# This function calculates heights of trees given distance of each tree 
# from its base and angle to its top, using  the trigonometric formula 
# height = distance * tan(radians)
# ARGUMENTS
# degrees:   The angle of elevation of tree
# distance:  The distance from base of tree (e.g., meters)
# OUTPUT
# The heights of the tree, same units as "distance"
TreeHeight <- function(degrees, distance){
  radians <- degrees * pi / 180
  height <- distance * tan(radians)
  #print(paste("Tree height is:", height))
  
  return (height)
}

# Test TreeHeight
# TreeHeight(37, 40)

###########################################################################
#                             Script
###########################################################################
# Read in file from the command line. 
# trailing only gets just the arguments passed after the script! (s)
inpath = commandArgs(trailingOnly = TRUE)
if (length(inpath) <1){
  print("No file provided as input! Using ../Data/trees.csv as default.")
  inpath = "../Data/trees.csv"
} else if (length(inpath) >1){
  print(paste("Script takes only 1 command line argument,", length(inpath), "given. Using the first argument as the file.", inpath[1]))
  inpath = inpath[1]
} else {
  print(paste("Reading from", inpath, sep = " "))
}

# strip of the file path and extension to get just the filename.
input = unlist(strsplit(inpath, "/"))[-1][-1]
output = gsub(".csv", "", input)

# format the input name as part of the output path to save to.
outpath = paste0("../Results/", output, "_treeheights.csv")

# read in the command line argument.
trees <- read.csv(inpath, header = TRUE)

# make a new column in trees with the value of the output of the TreeHeight function.
print("Calculating Tree heights...")
trees$Tree.Height.m <- TreeHeight(trees$Angle.degrees, trees$Distance.m)

print(paste("Analysis complete. File saved as ", outpath, sep = ""))
write.csv(trees, outpath, col.names = TRUE, row.names = FALSE)

File: Week3/Code/get_TreeHeight.py
#!/usr/bin/env python3
"""Calculates the height of a tree based on distance and angle."""
__appname__ = "get_TreeHeight.py"
__author__ = "Joseph Palmer <joseph.palmer18@imperial.ac.uk>"
__version__ = "0.0.1"
__license__ = "License for this code/"
__date__ = "Oct-2018"

## imports ##
import sys
import math
import csv
from pprint import pprint

## Functions ##

def TreeHeight(degrees, distance):
    """TreeHeight - Caluclates the hight of a tree based on
    the distance from the tree and the angle.

    :param degrees: float
    :param distance: float
    """
    radians = degrees * math.pi / 180
    height = distance * math.tan(radians)
    # print("The tree height is {}".format(height))
    return height

def GetFile():
    """GetFile - Open fiel given as command line argument.
    uses default if not given.
    """
    # set input file to work from. Open the file and read in as list.
    # have default in case one not provided.
    default_file = "../Data/trees.csv"
    if len(sys.argv) < 2:
        print("No command line arguments provided.\n"
              "using {} as default".format(default_file))
        filepath = default_file
    elif len(sys.argv) > 2:
        print("More than 1 command line argument provided.\n"
              "Using first argument: {}".format(sys.argv[1]))
        filepath = sys.argv[1]
    else:
        filepath = sys.argv[1]
    return filepath

def ReadCSV(filepath):
    """ReadCSV - Reads a file given and saves the output to a list of lines.

    :param filepath: str
    """
    f = open(filepath, "r")
    reader = csv.reader(f)
    reader = [i for i in reader]
    f.close()
    return reader


def main(argv):
    """main

    :param argv:
    """

    # get the file
    input_path = GetFile()
    
    # Read csv file into a list.
    reader = ReadCSV(input_path)
    
    # Define the headers and content to be calculated.
    headers = reader[:1][0]
    content = reader[1:]

    # Add the headers to the top of the results string.
    headers.append("Tree.Height.m")
    results_str = ",".join(headers) + "\n"

    # Extract the angle and distance values and feed into TreeHeight.
    # Append the tree height result to the original content as a new column
    # Format the content list of that line as a string and add to results_str
    print("Calculating tree height")
    for i in range(len(content)):
        angle = float(content[i][1])
        dis = float(content[i][2])
        content[i].append(str(TreeHeight(angle, dis)))
        results_str += ",".join(content[i]) + "\n"

    # get the location of where to save the output
    outname = input_path.split("/")[-1].replace(".csv", "")
    outpath = "../Results/{}_treeheights.csv".format(outname)
    print("Saving results to {}".format(outpath))

    # write the comma sepperated string to a new file
    with open(outpath, "w") as w:
        w.write(results_str)
    return 0


if __name__ == "__main__":
    # run if executed as script.
    main(sys.argv)
